File: 백준알고리즘/DFS(깊이 우선 탐색)/python/섬의 개수.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

w, h = map(int, input().split())
dx = [-1, 1, 0, 0, -1, 1, -1, 1]
dy = [0, 0, -1, 1, 1, -1, -1, 1]
cnt = 0


def dfs(x, y, graph):
    graph[x][y] = 0
    for i in range(8):
        nx = x + dx[i]
        ny = y + dy[i]

        if nx < 0 or nx >= h or ny < 0 or ny >= w:
            continue

        if graph[nx][ny] == 1:
            dfs(nx, ny, graph)


def solution(w, h, graph):
    global cnt
    for i in range(h):
        for j in range(w):
            if graph[i][j] == 1:
                cnt += 1
                dfs(i, j, graph)

    return cnt


while w != 0 and h != 0:
    graph = []
    for _ in range(h):
        graph.append(list(map(int, input().split())))

    print(solution(w, h, graph))
    graph = []
    w, h = map(int, input().split())
    cnt = 0

File: 백준알고리즘/DFS(깊이 우선 탐색)/javascript/섬의 개수.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().split('\n');
const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

let w;
let h;

while (w !== 0 && h !== 0) {
  [w, h] = input()
    .split(' ')
    .map(n => +n);
  if (w === 0 && h === 0) break;
  const graph = [];

  for (let i = 0; i < h; i++) {
    graph.push(
      input()
        .split(' ')
        .map(n => +n)
    );
  }
  const dx = [-1, 1, -1, 1, -1, 1, 0, 0];
  const dy = [0, 0, 1, -1, -1, 1, 1, -1];
  let cnt = 0;

  function dfs(x, y) {
    graph[x][y] = 0;

    for (let i = 0; i < 8; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || nx >= h || ny < 0 || ny >= w) continue;

      if (graph[nx][ny] === 1) dfs(nx, ny);
    }
  }

  for (let i = 0; i < h; i++) {
    for (let j = 0; j < w; j++) {
      if (graph[i][j] === 1) {
        cnt++;
        dfs(i, j);
      }
    }
  }

  console.log(cnt);
}
